clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java

```diff
@@ -1846,7 +1849,34 @@ public void testPollThrowsInterruptExceptionIfInterrupted() {
         }
         assertDoesNotThrow(() -> consumer.poll(Duration.ZERO));
     }
-    
+
+    @Test
+    void testReaperExpiresExpiredEvents() {
+        consumer = newConsumer();
+        final String topicName = "foo";
+        final int partition = 3;
+        final TopicPartition tp = new TopicPartition(topicName, partition);
+        final SortedSet<TopicPartition> partitions = new TreeSet<>(TOPIC_PARTITION_COMPARATOR);
+        partitions.add(tp);
+
+        consumer.subscribe(Collections.singletonList(topicName), new CounterConsumerRebalanceListener());
+
+        final ConsumerRebalanceListenerCallbackNeededEvent event1 = new ConsumerRebalanceListenerCallbackNeededEvent(ON_PARTITIONS_REVOKED, partitions);
+        final ConsumerRebalanceListenerCallbackNeededEvent event2 = new ConsumerRebalanceListenerCallbackNeededEvent(ON_PARTITIONS_ASSIGNED, partitions);
+        backgroundEventReaper.add(event1);
+        backgroundEventQueue.add(event2);
+
+        assertEquals(1, backgroundEventReaper.size());
+        assertEquals(1, backgroundEventQueue.size());
+
+        consumer.close();
+
+        assertEquals(0, backgroundEventReaper.size());
+        assertEquals(0, backgroundEventQueue.size());
+        assertTrue(event1.future().isCompletedExceptionally());
+        assertTrue(event2.future().isCompletedExceptionally());
+    }
```

Comment:
If you used a mock for the reaper, something like this would be enough to verify the correct use of the reaper in `close()`.

```suggestion
    void testReaperIsCalledInClose() {
        consumer = newConsumer();

        consumer.close();

        verify(backgroundEventReaper).reap(backgroundEventQueue)
    }
```

Reply:
Done.